src/transformers/models/mamba2/modeling_mamba2.py

```diff
@@ -0,0 +1,1123 @@
+            if not self.rms_norm:
+                gate = gate.view(batch_size, self.intermediate_size, self.head_dim)
+                # gate = rearrange(gate, "b (h p) -> b h p", p=self.head_dim)
+            hidden_states = selective_state_update(
+                cache_params.ssm_states[self.layer_idx],
+                hidden_states_reshaped,
+                dt,
+                A,
+                B,
+                C,
+                D,
+                z=gate if not self.rms_norm else None,
+                dt_bias=dt_bias,
+                dt_softplus=True,
+            )
+            hidden_states = hidden_states.view(batch_size, self.num_heads * self.head_dim)
+            if self.rms_norm:
+                hidden_states = self.norm(hidden_states, gate)
+            if d_mlp > 0:
+                hidden_states = torch.cat([torch.nn.functional.silu(z0) * x0, hidden_states], dim=-1)
+
+            out = self.out_proj(hidden_states).unsqueeze(1)
+        # if no cache is found, calling the kernel
+        else:
+            # 1. Gated MLP's linear projection
+            projected_states = self.in_proj(hidden_states)
+            A = -torch.exp(self.A_log.float())  # (num_heads) or (intermediate_size, state_size)
+            dt_limit_kwargs = {} if self.time_step_limit == (0.0, float("inf")) else {"dt_limit": self.time_step_limit}
+
+            if self.training and cache_params is None:
+                out, ssm_state = mamba_split_conv1d_scan_combined(
+                    projected_states,
+                    self.conv1d.weight.squeeze(1),
+                    self.conv1d.bias,
+                    self.dt_bias,
+                    A,
+                    D=self.D.view(-1, self.head_dim) if self.D_has_hdim else self.D,
+                    chunk_size=self.chunk_size,
+                    seq_idx=None,  # was seq_idx
+                    activation=self.activation,
+                    rmsnorm_weight=self.norm.weight,
+                    rmsnorm_eps=self.norm.variance_epsilon,
+                    outproj_weight=self.out_proj.weight,
+                    outproj_bias=self.out_proj.bias,
+                    headdim=None if self.D_has_hdim else self.head_dim,
+                    ngroups=self.n_groups,
+                    norm_before_gate=self.norm_before_gate,
+                    return_final_states=True,
+                    **dt_limit_kwargs,
+                )
+
+            else:
+                gate, xBC, time_step = torch.split(
+                    projected_states,
+                    [self.intermediate_size, self.conv_dim, self.num_heads],
+                    dim=-1,
+                )
+                if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
+                    # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
+                    # bug in generate tests?
```

Comment:
potentially because some test use 4d mask!

Reply:
I think I got rid of all generate bugs in the end, they were related to the attention mask slicing!

Reply:
cool!